### ISSUE TITLE
Add antialiased point-filtering

### ISSUE THREAD

**Issue by CptPotato** [godotengine/godot#20023]
So, a while ago in another engine, I've written a simple custom filtering algorithm that performs pretty much "perfect" antialiasing on point filtered textures under any viewing angle. Surprisingly, I haven't seen any game use this technique so far.

This could probably be written using normal Godot shaders, though it would be cumbersome to do so for all your objects (as somone over at Discord told me), so I thought I'd share it here.

![grafik](https://user-images.githubusercontent.com/3957610/42414006-af30b9fa-822c-11e8-96b3-be708cc8d4cc.png)

A few notes:
- it does not require a post processing pass as it is done in the object shader when sampling the texture
- it does not cause flickering, jittering, etc. of lines like FXAA, SMAA, etc. does
- it doesn't conflict with MSAA (it actually works well with it)
- I'm not sure how well it works with PBR maps
- It does require support for derivatives in pixel shaders; I'm not sure if this is supported on all devices
- just a thought: it could also be exposed as a custom sampling function for writing shaders in Godot (as an alternative to `texture(...)`)

Here's a more technical description (for shader guys):
It's basically using linear filtering with custom weights to make the gradient between the texture's pixels smaller. This together with pixel shader derivatives allows making this linear gradient 1 (screen) pixel wide, resulting in a look like supersamling with a very high sample count.
Lastly, you can simply turn the weights into offsets and use a linear sampler instead. This way you don't have to sample the 2x2 pixel quad yourself but let the HW acceleration do the job and only use a single sample.

Long story short: it's almost free.

Anyway, if you want to implement this, [here's the HLSL (d3d9) implementation](https://gist.github.com/CptPotato/c88454cb9473e76f535ac624c70779bf), it should be almost trivial to port over to GLSL.

For use with multiple textures you would want to do the maths only once and just do the sampling for every texture (assuming the texture sizes are the same).

**Comment by clayjohn** [comment]
Feature and improvement proposals for the Godot Engine are now being discussed and reviewed in a dedicated [**Godot Improvement Proposals (GIP)**](https://github.com/godotengine/godot-proposals) ([godotengine/godot-proposals](https://github.com/godotengine/godot-proposals)) issue tracker. The GIP tracker has a detailed issue template designed so that proposals include all the relevant information to start a productive discussion and help the community assess the validity of the proposal for the engine.

The main ([godotengine/godot](https://github.com/godotengine/godot)) tracker is now solely dedicated to bug reports and Pull Requests, enabling contributors to have a better focus on bug fixing work. Therefore, we are now closing all older feature proposals on the main issue tracker.

If you are interested in this feature proposal, please [**open a new proposal on the GIP tracker**](https://github.com/godotengine/godot-proposals/issues/new/choose) following the given issue template (after checking that it doesn't exist already). Be sure to reference this closed issue if it includes any relevant discussion (which you are also encouraged to summarize in the new proposal). Thanks in advance!
